[loaders] handle password-protected zip files

ajkerrigan opened the pull request:
If we wrap zipfile's `open()` method, we can catch and handle "password required" errors with VisiData's typical input flow.

This stems from a question/discussion with geekscrapy in the IRC channel. Some important details:

* `display=False` prevents the password from echoing back or being recorded in the command log for later replay.
* Leaving out a `type` for `vd.input()` prevents the password from being stored in a history buffer.

Thanks for the tips along the way @saulpw !

ajkerrigan committed:
[loaders] handle password-protected zip files

saulpw reviewed (changes requested):
I wonder whether it's better to inline it in `openRow` instead.  With monkey-patching like this, any VisiData plugin which uses `ZipFile.open` in a command will automatically use this same workflow, which is nice. But a non-VisiData tool that imports visidata and tries to open a password-protected .zip file will then also try to use `vd.input`, which won't work outside of `vd.run()`.

ajkerrigan committed:
Don't monkeypatch zipfile.open

Keep a zipfile.open wrapper method local to ZipSheet, to avoid
unintended side effects.

ajkerrigan commented:
> I wonder whether it's better to inline it in `openRow` instead. With monkey-patching like this, any VisiData plugin which uses `ZipFile.open` in a command will automatically use this same workflow, which is nice. But a non-VisiData tool that imports visidata and tries to open a password-protected .zip file will then also try to use `vd.input`, which won't work outside of `vd.run()`.

That's a good call, I wasn't sure where to slot this functionality myself. I _had_ to use the monkeypatching wrapper for local testing in `.visidatarc`, but I see the point about avoiding surprises elsewhere. Here's another option that avoids throwing too much logic directly into `openRow`, but also keeps any additions local to `ZipSheet`.

saulpw commented on visidata/loaders/archive.py:
```diff
@@ -16,9 +16,19 @@ class ZipSheet(Sheet):
         ColumnAttr('compress_size', type=int)
     ]
 
+    def open_zipfile(self, fp, *args, **kwargs):
```
We should use something other than `open_`, unless we want `-f zipfile` to work somehow (and then this would have to return a sheet of some kind).  Maybe `openZipFile`?

saulpw commented on visidata/loaders/archive.py:
```diff
@@ -16,9 +16,19 @@ class ZipSheet(Sheet):
+        except RuntimeError as err:
+            if 'password required' in err.args[0]:
+                pwd = vd.input(f'{args[0].filename} is encrypted, enter password: ', display=False)
+                return fp.open(*args, **kwargs, pwd=pwd.encode())
```
Let's make `'utf-8'` an explict parameter to `encode()`.  (It's the default option but would be helpful I think, especially given that this is a password.)

ajkerrigan committed:
Use explicit encoding, avoid open_* naming

saulpw reviewed (approved):
looks good, let's ship it! :ship:
